engine/src/main/java/org/hibernate/search/backend/impl/DeleteByQuerySupport.java

```diff
@@ -41,221 +29,55 @@ private DeleteByQuerySupport() {
-						}
-					}
-					else {
-						Type type = query.getType();
-						BytesRef valueAsBytes;
-						switch ( type ) {
-							case INT:
-							case FLOAT: {
-								int value;
-								if ( type == Type.FLOAT ) {
-									value = NumericUtils.floatToSortableInt( (Float) query.getValue() );
-								}
-								else {
-									value = (Integer) query.getValue();
-								}
-								BytesRefBuilder builder = new BytesRefBuilder();
-								NumericUtils.intToPrefixCoded( value, 0, builder );
-								valueAsBytes = builder.get();
-								break;
-							}
-							case LONG:
-							case DOUBLE: {
-								long value;
-								if ( type == Type.DOUBLE ) {
-									value = NumericUtils.doubleToSortableLong( (Double) query.getValue() );
-								}
-								else {
-									value = (Long) query.getValue();
-								}
-								BytesRefBuilder builder = new BytesRefBuilder();
-								NumericUtils.longToPrefixCoded( value, 0, builder );
-								valueAsBytes = builder.get();
-								break;
-							}
-							default:
-								throw new AssertionFailure( "has to be a Numeric Type at this point!" );
-						}
-						return new TermQuery( new Term( query.getFieldName(), valueAsBytes ) );
+					map.put( SingularTermDeletionQuery.QUERY_KEY, new SingularTermDeletionQueryMapper() );
+
+					DeletionQueryMapper[] arr = new DeletionQueryMapper[map.size()];
+					for ( Map.Entry<Integer, DeletionQueryMapper> entry : map.entrySet() ) {
+						arr[entry.getKey()] = entry.getValue();
 					}
 
+					MAPPERS = arr;
 				}
-
-			} );
-
-			TO_LUCENE_QUERY_CONVERTER = new ToLuceneQuery[map.size()];
-			for ( Map.Entry<Integer, ToLuceneQuery> entry : map.entrySet() ) {
-				TO_LUCENE_QUERY_CONVERTER[entry.getKey()] = entry.getValue();
+			}
+			finally {
+				MAPPERS_LOCK.unlock();
 			}
 		}
+		return MAPPERS[queryKey];
 	}
 
-	public static ToLuceneQuery getToLuceneQuery(int queryKey) {
-		return TO_LUCENE_QUERY_CONVERTER[queryKey];
-	}
-
-	private static final Set<Class<? extends DeletionQuery>> SUPPORTED_TYPES;
-	static {
-		{
-			Map<Integer, Class<? extends DeletionQuery>> map = new HashMap<>();
-
-			map.put( SingularTermQuery.QUERY_KEY, SingularTermQuery.class );
-
-			SUPPORTED_TYPES = Collections.unmodifiableSet( new HashSet<>( map.values() ) );
-		}
-	}
+	private static final Lock SUPPORTED_TYPES_LOCK = new ReentrantLock();
+	private static Set<Class<? extends DeletionQuery>> SUPPORTED_TYPES;
 
 	public static boolean isSupported(Class<? extends DeletionQuery> type) {
-		return SUPPORTED_TYPES.contains( type );
-	}
-
-	private static final StringToQueryMapper[] FROM_STRING;
-	static {
-		{
-			Map<Integer, StringToQueryMapper> map = new HashMap<>();
-
-			map.put( SingularTermQuery.QUERY_KEY, new StringToQueryMapper() {
+		if ( SUPPORTED_TYPES == null ) {
+			//LAZY init, but don't to it more than once
+			SUPPORTED_TYPES_LOCK.lock();
+			try {
+				if ( SUPPORTED_TYPES == null ) {
+					Map<Integer, Class<? extends DeletionQuery>> map = new HashMap<>();
```

Comment:
I'm a bit lost here; Why is this map created? In the end only the value set seems to be relevant. Also is the lazy creation worth all the effort (locking etc.)? On first thought I'd say that a simple map should suffice for both the methods of this class:

    private static Map<Class<? extends DeletionQuery>, DeletionQueryMapper> MAPPERS_BY_QUERY=
        Collections.unmodifiableMap(
            Collections.singletonMap(SingularTermDeletionQuery.class, new SingularTermDeletionQueryMapper() )
        );

    public static boolean isSupported(Class<? extends DeletionQuery> type) {
        return MAPPERS_BY_QUERY.containsKey( type );
    }

    public static DeletionQueryMapper getMapper(Class<? extends DeletionQuery> type) {
        return MAPPERS_BY_QUERY.get( type );
    }

That'd be much simpler, but maybe I'm just missing some piece? One thing I'm unsure is how would alternative `DeletionQuery` implementations be registered? Is that supported at all?